### DISCUSSION TITLE
.NET Aspire: Scalar displays API instance url when accessing via proxy

### DISCUSSION THREAD

**Discussion by anvillus** [scalar/scalar#4025] [Q&A]
Hi,

I'm trying to configure Scalar on an ASP.NET Core API project (which is itself part of a .NET Aspire project). By default, .NET Aspire adds a proxy in front of my API instance. The proxy uses the url defined by `applicationUrl` in my API's `launchSettings.json` (`localhost:8080`) while my API instance is assigned a random port by Aspire.

So when I run my Aspire application, the Aspire dashboard shows my API available at `localhost:8080` (the proxy):

![image](https://github.com/user-attachments/assets/c4c04ad2-9b03-4b07-bad9-a140ea61d064)

While the actual instance of my API is available at `localhost:38181`:

![image](https://github.com/user-attachments/assets/302c4efc-fcaf-461d-a581-0b104464ecfe)

But then when I navigate to `localhost:8080/scalar/v1`, everything looks fine, except the server url is that of the instance:

![image](https://github.com/user-attachments/assets/4ba87f0f-ab24-4ce6-8419-9f756109b193)

And if I try to use the `Test request` feature, sure enough I get a CORS error.

Now I understand that I'm seeing the instance url as the server url because `localhost:8080` proxies my request to `scalar/v1` to the instance at `localhost:38181`.
But I tried the same setup using Swagger instead of Scalar, and Swagger does set the API url to the base url of whatever url I used to access it from my browser (either the proxy or the instance). And that's because Swagger uses a [middleware](https://github.com/domaindrivendev/Swashbuckle.AspNetCore/blob/69db0e37bd27998c395a6200b830e42333d03171/src/Swashbuckle.AspNetCore.SwaggerUI/SwaggerUIMiddleware.cs#L84-L90) which checks, upon each request, whether the requested resource is the Swagger UI, and if so, "generates" the Swagger UI file using url available from the `HttpContext`. While Scalar, it appears, sets the server url at startup using the url of the instance.

Ideally, I would like to be able to replicate Swagger's behavior: if I access Scalar from the proxy at `localhost:8080`, I want that to be the server url. If I access Scalar from `localhost:xxxx`, I want that to be the server url.

I have tried several things:
* Configuring the CORS on my API to allow the proxy url. That does solve the CORS issue mentioned above but not the discrepancy between the url in my browser/the Aspire dashboard and the server url.
* Passing the proxy url to `options.AddServer` or `options.Servers`. This does indeed replace the instance url with the proxy url but then the problem is the other way around: if I access my instance at `localhost:38181` then the server url is that of the proxy and I'm back to square one with CORS issues.
* Configuring Scalar from a custom middleware by calling `MapScalarApiReference` from the middleware instead of at startup (using the same configuration as in the previous bullet point). This simply doesn't work, `scalar/v1` returns a 404. I also tried to call both at startup (without any configuration) and from the middleware (setting `options.Servers` or `AddServer` with the url obtained from the `HttpContext`), the latter call has no effect: I still see the url of the instance even though `options.Servers` does contain the url of the proxy.
* Disabling the proxy set by Aspire. This obviously works but deprives me of the possibility to create multiple instances of my API should I ever want to.

So I'm at a loss here, I'm not even sure if it's more of a Scalar issue or an Aspire issue. I did find a similar issue [issue](https://github.com/dotnet/aspire/issues/1507) on the Aspire repository but the Aspire team said it was by design, what with the whole proxy thing.

**Comment by xC0dex** [comment]
Hey @lacheteau-s,

Thanks for opening a discussion. The issue that you are experiencing is related to the current implementation of the `Microsoft.AspNetCore.OpenApi` package. [This issue](https://github.com/dotnet/aspnetcore/issues/57332) explains the problem that you are currently experiencing. Scalar only uses the server information that is provided in the generated document. 

Please check out [this repo](https://github.com/xC0dex/AspireScalar), maybe the code is interesting for you. [This line](https://github.com/xC0dex/AspireScalar/blob/main/AspireScalar.ApiService/Program.cs#L9) adds a document transformer which dynamically fetches the host and replaces the server list in the OpenAPI document. [This](https://github.com/xC0dex/APIWeaver/blob/main/src/APIWeaver.OpenApi/Extensions/OpenApiOptionsExtensions.cs#L190) is the implementation of the document transformer.

**Edit:** There are also some workarounds described in the referenced issue.

**Reply by anvillus** [reply]
Thanks for your reply @xC0dex, I'm gonna give it a shot.

In the meantime, I also found that the following solves the issue somehow:

```
app.MapScalarApiReference(options =>
{
    options.Servers = Array.Empty<ScalarServer>();
});
```

By default `Servers` is `null`. And I don't understand why setting it to an empty list versus `null` would make a difference as to what url is being used by the API client. I browsed [Scalar.AspNetCore](https://github.com/scalar/scalar/tree/main/packages/scalar.aspnetcore/src/Scalar.AspNetCore) to try and make sense out of it but I couldn't find any reference to `Servers` other than the declaration and the `AddServer` extension method. So my guess is that it all happens in the JS code when [serving the page](https://github.com/scalar/scalar/blob/429e4f294fa03ee00554fefcdf47d184e8820bfa/packages/scalar.aspnetcore/src/Scalar.AspNetCore/Extensions/ScalarEndpointRouteBuilderExtensions.cs#L55-L88). I'm not gonna dig that deep but could you maybe shed some light on this?

**Reply by xC0dex** [reply]
Hey @lacheteau-s, great that you found a workaround. By default, Scalar uses the server from the generated OpenAPI document. However, you can overwrite these values with the `Servers` property. This is exactly what you do in your workaround. I assume that the API reference uses the current `window.location` if no servers were provided.
